Increase granularity of activity tracking to per turn.

kevingranade opened the pull request:
#### Summary
SUMMARY: None

#### Purpose of change
Increase activity tracking granularity to per turn.
Simplify character class slightly by extracting activity and weariness logic to a helper class.

#### Describe the solution
Extract the weariness and activity tracking code into a separate class.
Also activity level was not being saved or loaded, do that.
Increase granularity of tracking to per-turn.
Maintains a rolling average of activity level for bmr calculation, and also an instantaneous activity level for other uses.

#### Describe alternatives you've considered
I'm not at all happy with the names, activity_tracker is okish, but activity_history is just bad.

#### Testing
Add a number of unit tests, insure the existing weariness and calorie expenditure tests continue to work.
Run around and do active stuff manually and validate the results.

anothersimulacrum commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,46 @@
+        void deserialize( JsonIn &jsin );
+};
+
+#endif /* CATA_SRC_ACTIVITY_TRACKER_H */
```
Looks like it wants
```suggestion
#endif // CATA_SRC_ACTIVITY_TRACKER_H
```

kevingranade commented:
Updated with WIP adjustments to make activity ticks more granular by tracking them on a turn-by-turn basis and then applying them once every 5 minutes.

I'm kind of stuck what this does that breaks weariness so badly, it absolutely DESTROYS the weariness tests because weariness passes an inflection point and then just keeps climbing rapidly, I'm not sure what I did to make this happen.

kevingranade commented on tests/activity_tracker_test.cpp:
```diff
@@ -0,0 +1,97 @@
+    int end_value = values.back();
+    // activity() still returns most recently logged activity.
+    CHECK( tracker.activity() == end_value );
+    const float expected_activity = std::accumulate( values.begin(), values.end(), 0 );
```
```suggestion
    const float expected_activity = std::accumulate( values.begin(), values.end(), 0.0f );
```

kevingranade commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,53 @@
+        // outputs player activity level to a printable string
+        std::string activity_level_str() const;
+        float activity() const;
+        float total_activity() const;
```
```suggestion
```

kevingranade commented on tests/activity_tracker_test.cpp:
```diff
@@ -0,0 +1,98 @@
+    int end_value = values.back();
+    // activity() still returns most recently logged activity.
+    CHECK( tracker.activity() == end_value );
+    const float expected_activity = std::accumulate( values.begin(), values.end(), 0.0f ) /
+        static_cast<float>( values.size() );
```
```suggestion
    const float expected_activity = std::accumulate( values.begin(), values.end(), 0.0f ) /
                                   static_cast<float>( values.size() );
```

kevingranade commented on tests/activity_tracker_test.cpp:
```diff
@@ -0,0 +1,98 @@
+    // activity() still returns most recently logged activity.
+    CHECK( tracker.activity() == end_value );
+    const float expected_activity = std::accumulate( values.begin(), values.end(), 0.0f ) /
+                                   static_cast<float>( values.size() );
```
```suggestion
                                    static_cast<float>( values.size() );
```

AlexMooney commented on src/character.h:
```diff
@@ -2041,10 +2028,10 @@ class Character : public Creature, public visitable<Character>
         // does not decrease activity level
-        void increase_activity_level( float new_level );
-        // decreases the activity level to the previous level
+        void set_activity_level( float new_level );
```
I don't think this is quite the right name.  How about `log_activity_level`?

kevingranade replied:
log_activity_level() is already taken.

kevingranade commented on src/activity_tracker.cpp:
```diff
@@ -0,0 +1,136 @@
+void activity_tracker::try_reduce_weariness( int bmr, bool sleeping )
+{
+    tick_counter++;
+    if( average_activity() - NO_EXERCISE < 0.1 ) {
```
There's a pretty big difference between 0.1 "practically no exercise" and 0.5 "equal parts no exercise and light exercise".
I can avoid a magic number though, I'll replace it with an official epsilon.

kevingranade replied:
There's a pretty big difference between 0.1 "practically no exercise" and 0.5 "equal parts no exercise and light exercise".
I can avoid a magic number though, I'll replace it with an official epsilon.

actual-nh commented on src/activity_tracker.cpp:
```diff
@@ -0,0 +1,136 @@
+    tick_counter = 0;
+}
+
+std::string activity_tracker::debug_weary_info() const
```
Good idea! Note that this information would also be useful in weary_tests.cpp (tracked with each change of weariness level, as weariness and weary_threshold are now).

anothersimulacrum commented on src/activity_tracker.cpp:
```diff
@@ -0,0 +1,136 @@
+        accumulated_activity = 0.0f;
+        num_events = 1;
+    } else {
+        accumulated_activity += current_activity;
+        // skipped turns?
+        current_activity = new_level;
+        num_events++;
```
I think this should look like this to account for the turns thing
```suggestion
        int turns_moved = calendar::turn - current_turn;
        accumulated_activity += current_activity * turns_moved;
        // skipped turns?
        current_activity = new_level;
        num_events += turns_moved;
```

kevingranade replied:
Sort of, we can do number of turns, but the activity level if anything should be NO_EXERCISE, representative of the fact that no activity was logged in the interim.

anothersimulacrum replied:
I'd have to look through to see when skipped turns trigger, but this is the correct behavior for what I was considering, when an activity takes more than a characters moves to complete. E.g. if I drag furniture and it takes me 6 turns (but it is not an activity), I should be counted as whatever activity level for 6 turns, instead of just 1, and nothing for 5 turns.

kevingranade replied:
The activity should be logging its exertion level every turn.

anothersimulacrum replied:
I am talking about non-activity actions like that. Dragging is now an activity, but if it were not this would be the correct behavior.
I am not sure how many of these we have, however. Melee combat comes to mind as somewhere this could pop up.

kevingranade replied:
Still, anything the player does should log an activity, and anything that takes more than one turn should log each turn.
If that isn't happening, we can fix it.

kevingranade commented on tests/stomach_contents_test.cpp:
```diff
@@ -94,6 +94,8 @@ TEST_CASE( "starve_test", "[starve][slow]" )
     Character &dummy = get_player_character();
     reset_time();
     clear_stomach( dummy );
+    dummy.reset_activity_level();
```
```suggestion
    dummy.reset_activity_level();
    calendar::turn += 1;
```

kevingranade commented on tests/stomach_contents_test.cpp:
```diff
@@ -94,6 +94,9 @@ TEST_CASE( "starve_test", "[starve][slow]" )
     reset_time();
     clear_stomach( dummy );
+    dummy.reset_activity_level();
+    calendar::turn += 1;
```
```suggestion
    calendar::turn += 1_seconds;
```

kevingranade commented:
Rebased and fixed the failing starve test. It was being thrown off by the rolling average of activity level, so if another test happened right before it, it was picking up an unusually high activity level.

kevingranade commented:
Rebased on latest master and should be ready to go, I incorporated @anothersimulacrum 's activity level reporting feature into the activity tracker.

For future consideration, I want to unify move point, stamina, and weariness tracking behind a method called Character::expend_effort( int moves, float exertion_level )  That way we never expend moves without logging effort, and we never log effort without expending moves, and the product of moves spent and exertion level should be sufficient to determine stamina and fatigue for the action as well as being the sole place to spend moves.

actual-nh commented:
> For future consideration, I want to unify move point, stamina, and weariness tracking behind a method called Character::expend_effort( int moves, float exertion_level ) That way we never expend moves without logging effort, and we never log effort without expending moves, and the product of moves spent and exertion level should be sufficient to determine stamina and fatigue for the action as well as being the sole place to spend moves.

Spending of calories?

actual-nh commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,54 @@
+#ifndef CATA_SRC_ACTIVITY_TRACKER_H
+#define CATA_SRC_ACTIVITY_TRACKER_H
+
+#include "calendar.h"
```
```suggestion
#include <limits>

#include "calendar.h"
```

kevingranade commented:
> Spending of calories?

Yes, exertion level would feed into calorie expenditure.

actual-nh commented:
This PR has some of the same thinking as #47653, namely trying to smooth out some of the current fluctuations due to things happening every X minutes. If it goes in prior to #47653, it will unfortunately require a bit of a rewrite on my part of #47653 - or vice-versa, if #47653 goes in first - but that would just be an annoyance.

actual-nh commented:
BTW, the two failing builds are both at the `activity_counter_from_1_to_300` test; tracker.average_activity() vs expected_activity  1.00863f vs 150.5f. (The Travis one is for mingw-w64.)

anothersimulacrum commented:
I've assigned myself to this so I (hopefully) remember to review it.

kevingranade commented:
I THINK what's happening with that failing test is that calendar::turn is relatively advanced, and the loop in activity_tracker::new_turn() is incorporating the difference between calendar::turn_zero and calendar::turn into the tracker, resulting in the activity score getting greatly depressed.  The thing I don't get is why it isn't always happening.

actual-nh commented on tests/activity_scheduling_helper.cpp:
```diff
@@ -103,14 +103,14 @@ weariness_events do_activity( tasklist tasks )
         while( turns <= task.interval && !task.instantaneous() ) {
-            // Start each turn with a fresh set of moves
-            guy.moves = 100;
-            task.do_turn( guy );
             // Advance a turn
             calendar::turn += 1_turns;
             turns += 1_seconds;
             // Consume food, become weary, etc
             guy.update_body();
+            // Start each turn with a fresh set of moves
+            guy.moves = 100;
+            task.do_turn( guy );
         }
```
I can see why putting the do_turn after advancing turns makes sense. But why update body prior to do_turn?

kevingranade replied:
Because that's the order in which it occurs in game::do_turn().
Technically moves are incremented at the end of the turn instead of the start, but update_body() shouldn't be interacting with moves, so it shouldn't matter.

AlexMooney commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,54 @@
+        float activity() const;
+        // Returns average of activity level for the current period.
+        float average_activity() const;
+        // Returns the previous turn's activity level until an action is tanken on the current turn.
```
```suggestion
        // Returns the previous turn's activity level until an action is taken on the current turn.
```

AlexMooney commented on src/character.cpp:
```diff
@@ -5651,7 +5635,7 @@ std::pair<int, int> Character::weariness_transition_progress() const
 int Character::weariness_level() const
 {
-    int amount = weariness();
+    int amount = activity_history.weariness();
```
Seems like with the change to `Character::weariness()` this could be left alone.

ZhilkinSerg replied:
```suggestion
    int amount = weariness();
```

AlexMooney commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,54 @@
+        void new_turn();
+        // Resets accumulated activity level.
+        void reset_activity_level();
+        // outputs player activity level to a printable string
```
```suggestion
        // Outputs player activity level to a printable string.
```

AlexMooney commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,54 @@
+        int intake = 0;
+        // Semi-consecutive 5 minute ticks of low activity (or 2.5 if we're sleeping)
+        int low_activity_ticks = 0;
+        // How many ticks since we've decreased intake
```
```suggestion
        // How many ticks since we've decreased intake.
```

AlexMooney commented on src/activity_tracker.h:
```diff
@@ -0,0 +1,54 @@
+        // Weariness metadata.
+        int tracker = 0;
+        int intake = 0;
+        // Semi-consecutive 5 minute ticks of low activity (or 2.5 if we're sleeping)
```
```suggestion
        // Semi-consecutive 5 minute ticks of low activity (or 2.5 if we're sleeping).
```
